### ISSUE TITLE
Allow overriding of hard-coded `postgres` database in public schema configuration and improve error message when `postgres` database does not exist

### ISSUE THREAD

**Issue by tjjjwxzq** [rails/rails#48236]
### Context
In the event that you need to run rails db tasks against a postgresql server that does not have a `postgres` database, `rails db:drop` and `db:create` and any other db tasks that depend on them will fail and complain with misleading and confusing error messages. 

The root cause is the database `postgres` being hardcoded in the connection configuration: https://github.com/rails/rails/blob/d137b10f946ff78fac8fe203d5aeaf2bb4c3a1d9/activerecord/lib/active_record/tasks/postgresql_database_tasks.rb#L102-L104

when a `drop` or `create` is being attempted, but the error messages shown, especially for the `drop` command, are misleading and confusing since they either suggest a database that is actually present does not exist, or that the fault lies in the user's `config/database.yml` file, when neither is the case.

A proposal to fix this behaviour is:

1. When the `ActiveRecord::NoDatabaseError` is thrown because the connection to the `postgres` database fails, and not because the connection to the user's specified database to drop/create fails, improve the error message being shown to the user (this [line](https://github.com/rails/rails/blob/d137b10f946ff78fac8fe203d5aeaf2bb4c3a1d9/activerecord/lib/active_record/tasks/database_tasks.rb#L121-L124) for `create` and this [line](https://github.com/rails/rails/blob/d137b10f946ff78fac8fe203d5aeaf2bb4c3a1d9/activerecord/lib/active_record/tasks/database_tasks.rb#L208-L209) for `drop`)
 a. For `create`, the `ActiveRecord::NoDatabaseError` message ends up showing something like `We could not find your database: postgres. Which can be found in the database configuration file located at config/database.yml.` which is confusing to the user as the problem is not with anything in `config/database.yml`, but with the fact that the `postgres` database which rails assumes always exists does not exist
b. For `drop`, the messaage shown after `ActiveRecord::NoDatabaseError` is rescued misleads the user into thinking that the database that they tried to drop doesn't exist, when the error is actually raised because the code tries to connect to a non-existent `postgres` database
2. Allow users to override the default database these tasks connect to instead of hardcoding `postgres`. The improved error messages can then prompt the user to either 1) ensure that the `postgres` database exists, or 2) override the default database name when running those tasks. 

### Steps to reproduce for `db:drop`

Set up a PostgreSQL database that does not have a default `postgres` database. This can be done by explicitly dropping the `postgres` database with SQL commands, or using database providers such as DigitalOcean who do not create a `postgres` database by default (their default database is named `defaultdb`).

Attempt to run `./bin/rails db:drop` passing in the appropriate `DATABASE_URL`, say with a database called `my-db`. 

### Expected behavior
Complain that the `postgres` database does not exist and prompt user to create it or override the default database that the task connects to to carry out the drop operation on the specified database.

### Actual behaviour
This will complain: `Databse 'my-db' does not exist` and the database will fail to be dropped 

### Steps to reproduce for `db:create`
Attempt to run `./bin/rails db:create`, passing in the appropriate `DATABASE_URL`, say with a database called `my-db`. 

### Expected behavior
Complain that the `postgres` database does not exist and prompt user to create it or override the default database that the task connects to to carry out the create operation for the specified database, instead of suggesting that there is an error in the users `config/database.yml`

### Actual behaviour
This will throw an error:

```
We could not find your database: postgres. Which can be found in the database configuration file located at config/database.yml.

To resolve this issue:

- Did you create the database for this app, or delete it? You may need to create your database.
- Has the database name changed? Check your database.yml config has the correct database name.

To create your database, run:

        bin/rails db:create
Couldn't create 'my-db database. Please check your configuration.
rails aborted!
```

### System configuration
**Rails version**: 7.0.2.3

**Ruby version**: 3.1.1

**Comment by hahmed** [comment]
It looks like https://github.com/rails/rails/pull/47028 addresses this very issue 🤔, it's not released yet, but worth testing against the latest version of rails on main?

**Comment by tjjjwxzq** [comment]
Thanks @hahmed ! That does look like it should address the misleading error messages, I'll give it a try when I have the time. But perhaps we should also be allowing the user to override the assumed default `postgres` database?

**Comment by hahmed** [comment]
I think that makes sense, btw how do we handle the scenario where let's say a new rails app is created and the default db is not `postgres` but something else, the app developer will not have updated the config/override this error message will still be misleading?

**Comment by tjjjwxzq** [comment]
Cool, I'll try to work on a PR. I'm not sure why there would be a difference for a completely fresh setup where the default `postgres` db doesn't exist? In that case, the error message should just alert the user to the fact that the default `postgres` db that Rails is trying to connect to is missing, and prompt the user to either create the `postgres` db, or override the default db name. 

From what I can tell just reading the code and not having tested it, after #47028 , the error message should complain that rails can't connect to the `postgres` db, and at least not mislead the user into believing that the database they are trying to create/drop doesn't exist and there's something wrong with their `database.yml` config. But it could be improved further by adding the prompts I mentioned above.

**Comment by zzak** [comment]
From what I can tell, the new error message is not entirely inaccurate. I had to manually drop my "postgres" db in order to reproduce it:

```
PG::ConnectionBad: connection to server at "::1", port 5432 failed: FATAL:  database "postgres" does not exist
```

So technically it doesn't exist, which is correct.

> prompt user to create it or override the default database

I'm weary about adding too much implicit context to the error messages here, but perhaps documenting this process would be helpful. e.g. "how to override the default when this error occurs" if it's not already.

**Comment by hahmed** [comment]
I'm not sure if there is a way to override the default message at this moment 🤔

**Comment by asmrtfm** [comment]
Doesn't seem like this was ever resolved...

For example, if using digitalocean's postgres offerings, they automatically create a default postgres database named "defauldb" instead of "postgres" and it can not be changed.

Is there a config option that Rails respects for setting the name of the postgres default database?

So far the only way I've been able to get it to work is by setting the primary database to point to our remote production one on digitalocean, which of coarse breaks the development environment.

**Comment by asmrtfm** [comment]
I've submitted a [PR](https://github.com/rails/rails/pull/55715) to rails branch 7-2-stable (should also work up to main as the relevant code has not changed) to address this closed-but-unresolved issue.

---
Sorry about the mentions clutter.  Seems that github has, or had, an open issue (https://github.com/isaacs/github/issues/1081) about this undesirable behavior.  I opened up a discussion about it on their new forum thing.
